### PR TITLE
utils/parser: migrate test-infra to testify

### LINKED ISSUE #26981: migrate test-infra to testify for `util/parser` pkg

### PR DESCRIPTION
**Pull request by w169q169**
### What problem does this PR solve?

Issue Number: close #26981  

### Release note

```release-note
None
```

### CODE REVIEW

File: util/parser/ast_test.go [hunk 1]
```diff
@@ -60,12 +52,14 @@ func (s *testASTSuite) TestSimpleCases(c *C) {
 		},
 	}
 
-	for _, t := range tests {
+	for _, testCase := range tests {
```
Comment (tisonkun) [inline util/parser/ast_test.go:55]:
```suggestion
	for _, test := range tests {
```

nit: if we range `tests`, we'd better name it `test`.

File: util/parser/parser_test.go [hunk 2]
```diff
@@ -40,10 +42,10 @@ func (s *testParserSuite) TestSpace(c *C) {
 		{1, "     1", "1"},
 		{2, "  1", "1"},
 	}
-	for _, test := range okTable {
-		rest, err := Space(test.Input, test.Times)
-		c.Assert(rest, Equals, test.Expected)
-		c.Assert(err, IsNil)
+	for _, testCase := range okTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:45]: ditto

File: util/parser/parser_test.go [hunk 3]
```diff
@@ -54,14 +56,15 @@ func (s *testParserSuite) TestSpace(c *C) {
 		{2, " 1"},
 	}
 
-	for _, test := range errTable {
-		rest, err := Space(test.Input, test.Times)
-		c.Assert(rest, Equals, test.Input)
-		c.Assert(err, NotNil)
+	for _, testCase := range errTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:59]: ditto

File: util/parser/ast_test.go [hunk 4]
```diff
@@ -17,22 +17,14 @@ package parser_test
...
-type testASTSuite struct {
-}
-
-func TestT(t *testing.T) {
-	TestingT(t)
-}
-
-func (s *testASTSuite) TestSimpleCases(c *C) {
+func TestSimpleCases(t *testing.T) {
 	tests := []struct {
```
Comment (tisonkun) [inline util/parser/ast_test.go:28]: can we run this test in parallel?

File: util/parser/parser_test.go [hunk 5]
```diff
@@ -29,7 +31,7 @@ func TestT(t *testing.T) {
 	TestingT(t)
 }
 
-func (s *testParserSuite) TestSpace(c *C) {
+func TestSpace(t *testing.T) {
 	okTable := []struct {
```
Comment (tisonkun) [inline util/parser/parser_test.go:35]: can we run this test in parallel?

File: util/parser/parser_test.go [hunk 6]
```diff
@@ -54,14 +56,15 @@ func (s *testParserSuite) TestSpace(c *C) {
...
+	for _, testCase := range errTable {
+		rest, err := Space(testCase.Input, testCase.Times)
+
+		require.NotNil(t, err)
+		require.Equal(t, testCase.Input, rest)
 	}
 }
 
-func (s *testParserSuite) TestDigit(c *C) {
+func TestDigit(t *testing.T) {
```
Comment (tisonkun) [inline util/parser/parser_test.go:67]: can we run this test in parallel?

File: util/parser/parser_test.go [hunk 7]
```diff
@@ -74,11 +77,12 @@ func (s *testParserSuite) TestDigit(c *C) {
 		{3, "456 121", "456", " 121"},
 	}
 
-	for _, test := range okTable {
-		digits, rest, err := Digit(test.Input, test.Times)
-		c.Assert(digits, Equals, test.ExpectedDigits)
-		c.Assert(rest, Equals, test.ExpectedRest)
-		c.Assert(err, IsNil)
+	for _, testCase := range okTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:80]: you can keep `test` as golang tends to use short name

File: util/parser/parser_test.go [hunk 8]
```diff
@@ -90,15 +94,16 @@ func (s *testParserSuite) TestDigit(c *C) {
 		{3, "12 int"},
 	}
 
-	for _, test := range errTable {
-		digits, rest, err := Digit(test.Input, test.Times)
-		c.Assert(digits, Equals, "")
-		c.Assert(rest, Equals, test.Input)
-		c.Assert(err, NotNil)
+	for _, testCase := range errTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:97]: ditto

File: util/parser/parser_test.go [hunk 9]
```diff
@@ -90,15 +94,16 @@ func (s *testParserSuite) TestDigit(c *C) {
...
+		digits, rest, err := Digit(testCase.Input, testCase.Times)
+
+		require.NotNil(t, err)
+		require.Equal(t, "", digits)
+		require.Equal(t, testCase.Input, rest)
 	}
 }
 
-func (s *testParserSuite) TestNumber(c *C) {
+func TestNumber(t *testing.T) {
```
Comment (tisonkun) [inline util/parser/parser_test.go:106]: can we run this test in parallel?

File: util/parser/parser_test.go [hunk 10]
```diff
@@ -109,11 +114,12 @@ func (s *testParserSuite) TestNumber(c *C) {
 		{"123 @)@)", 123, " @)@)"},
 		{"456 121", 456, " 121"},
 	}
-	for _, test := range okTable {
-		digits, rest, err := Number(test.Input)
-		c.Assert(digits, Equals, test.ExpectedNum)
-		c.Assert(rest, Equals, test.ExpectedRest)
-		c.Assert(err, IsNil)
+	for _, testCase := range okTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:117]: ditto iter name

File: util/parser/parser_test.go [hunk 11]
```diff
@@ -124,15 +130,16 @@ func (s *testParserSuite) TestNumber(c *C) {
 		{"@)@)int"},
 	}
 
-	for _, test := range errTable {
-		digits, rest, err := Number(test.Input)
-		c.Assert(digits, Equals, 0)
-		c.Assert(rest, Equals, test.Input)
-		c.Assert(err, NotNil)
+	for _, testCase := range errTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:133]: ditto iter name

File: util/parser/parser_test.go [hunk 12]
```diff
@@ -124,15 +130,16 @@ func (s *testParserSuite) TestNumber(c *C) {
...
+		digits, rest, err := Number(testCase.Input)
+
+		require.NotNil(t, err)
+		require.Equal(t, 0, digits)
+		require.Equal(t, testCase.Input, rest)
 	}
 }
 
-func (s *testParserSuite) TestCharAndAnyChar(c *C) {
+func TestCharAndAnyChar(t *testing.T) {
```
Comment (tisonkun) [inline util/parser/parser_test.go:142]: ditto parallel

File: util/parser/parser_test.go [hunk 13]
```diff
@@ -143,14 +150,16 @@ func (s *testParserSuite) TestCharAndAnyChar(c *C) {
 		{'1', "12 int", "2 int"},
 	}
 
-	for _, test := range okTable {
-		rest, err := Char(test.Input, test.Char)
-		c.Assert(rest, Equals, test.Expected)
-		c.Assert(err, IsNil)
+	for _, testCase := range okTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:153]: ditto iter name

File: util/parser/parser_test.go [hunk 14]
```diff
@@ -162,9 +171,10 @@ func (s *testParserSuite) TestCharAndAnyChar(c *C) {
 		{'1', "x12 int"},
 	}
 
-	for _, test := range errTable {
-		rest, err := Char(test.Input, test.Char)
-		c.Assert(rest, Equals, test.Input)
-		c.Assert(err, NotNil)
+	for _, testCase := range errTable {
```
Comment (tisonkun) [inline util/parser/parser_test.go:174]: ditto iter name

**Review by tisonkun — COMMENTED** [review]
Thanks for your contribution @w169q169 ! Comments inline.

File: util/parser/ast_test.go [hunk 15]
```diff
@@ -17,22 +17,14 @@ package parser_test
 import (
 	"testing"
 
-	. "github.com/pingcap/check"
+	"github.com/stretchr/testify/require"
+
```
Comment (xiongjiwei) [inline util/parser/ast_test.go:21]: our import style is 
```go
import (
   stdlib

   third and self lib
)
```

Comment (xiongjiwei) [reply util/parser/ast_test.go:21]: please remove this empty line

**Comment by w169q169** [comment]
Thanks for your review @tisonkun @xiongjiwei. I already fix test code according to  opions, but I dont know why  ci is not passed。
Could you give me some help to fix the problem?

**Comment by tisonkun** [comment]
please download the log and grep "FAIL" case sensitive to see the root cause.

**Comment by unconsolable** [comment]
The goroutine leak created by initializeAsyncParsing has been fixed by #27471

Maybe you can eliminate the failure by updating base branch

**Comment by tisonkun** [comment]
/cc @xiongjiwei @xhebox 

Please take a look.

**Review by xhebox — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**